01_stratified_classification/nbconverted/plot_univariate_fs_results.py

```diff
@@ -0,0 +1,335 @@
+    ax = axarr[ix // 3, ix % 3]
+    plot_df = compare_df[compare_df.identifier == gene]
+    sns.boxplot(data=plot_df, x='fs_method', y='delta_aupr', ax=ax)
+    ax.set_title(gene)
+    ax.set_xlabel('Feature selection method')
+    ax.set_ylim(0, 1)
+
+plt.tight_layout()
+
+
+# We can see that the blue box (median f-test) and the purple box (pan-cancer f-test) are reasonably similar in most cases to the green box (no feature selection from 1000 features), suggesting that selecting features by f-statistic doesn't seem to harm performance, at the very least. This is good.
+
+# ### Further analysis of selected features
+# 
+# We want to look at:
+# * Overlap of features in at least one model (for different FS methods)
+# * f-statistic distributions for features in at least one model
+
+# In[6]:
+
+
+# gene to analyze features for
+gene = 'TP53'
+
+id_coefs_info = []
+for identifier, coefs_list in au.generate_nonzero_coefficients_fs(
+        results_dir, cfg.fs_methods):
+    if not identifier.startswith(gene): continue
+    for fold_no, coefs in enumerate(coefs_list):
+        id_coefs_info.append([identifier, fold_no, coefs])
+        
+print(len(id_coefs_info))
+
+
+# In[7]:
+
+
+print(len(id_coefs_info[0]))
+print(id_coefs_info[0])
+
+
+# In[8]:
+
+
+# list of sets, one for each feature selection method, of
+# features that were selected in at least one cross-validation fold
+fs_method_coefs = {}
+for coefs_list in id_coefs_info:
+    identifier = coefs_list[0]
+    features = list(zip(*coefs_list[2]))[0]
+    if identifier in fs_method_coefs:
+        fs_method_coefs[identifier].update(features)
+    else:
+        fs_method_coefs[identifier] = set(features)
+    
+print(list(fs_method_coefs.keys()))
+
+
+# In[9]:
+
+
+print(len(fs_method_coefs['{}_mad_n100'.format(gene)]))
+print(list(fs_method_coefs['{}_mad_n100'.format(gene)])[:5])
+
+
+# In[10]:
+
+
+def series_from_samples(samples, labels):
+    """Generate the weird dataframe format that Python upsetplot expects.
+    
+    Use as input lists of samples, and the labels that correspond
+    to each list.
+    """
+    # use pyvenn to generate overlaps/labels from sample IDs
+    venn_labels = generate_petal_labels(samples)
+    # generate format upset plot package expects
+    df_ix = [[(i == '1') for i in list(b)] + [int(v)] for b, v in venn_labels.items()]
+    # generate dataframe from list
+    rename_map = {ix: labels[ix] for ix in range(len(labels))}
+    index_names = list(rename_map.values())
+    rename_map[len(labels)] = 'id'
+    df = (pd.DataFrame(df_ix)
+        .rename(columns=rename_map)
+        .set_index(index_names)
+    )
+    # and return as series
+    return df['id']
+
+
+# In[11]:
+
+
+upset_series = series_from_samples(
+    list(fs_method_coefs.values()), list(fs_method_coefs.keys())
+)
+upset_series[upset_series != 0].sort_values(ascending=False).head(5)
+
+
+# In[12]:
+
+
+up.plot(upset_series[upset_series != 0])
```

Comment:
Making sure I'm keeping track of this. This is showing the number of cancer genes where TP53 is selected as a useful feature for prediction by all these various feature selection methods?

Reply:
Not quite, this is a bit confusing for sure - for models where we're predicting TP53 mutation status with various feature selection methods, this is showing the overlap in the gene expression features they selected (either the unique features for the single dots, or shared features for the connected dots in the upset plot). I'll add this to the Markdown cell too.